hutch_python/ipython_session_timer.py

```diff
@@ -0,0 +1,99 @@
+"""
+This module modifies an ``ipython`` shell to automatically close if it has been
+idle for 48 hours. Prior to starting hutch-python the user is alerted that the
+shell will automatically close after another 24 hours and is given the option
+to enter the number of hours they want to extend the session.
+"""
+
+import time
+from threading import Thread
+
+from IPython import get_ipython
+
+max_idle_time = 172800
+
+
+def configure_timeout(duration):
+    if isinstance(duration, int) and duration > 0:
+        max_idle_time = duration
+
+
+class IPythonSessionTimer:
+    '''
+    Class tracks the amount of time the current `InteractiveShell` instance (henceforth
+    called 'user session') has been idle and closes the session if more than 48
+    hours have passed.
+
+    Time is in seconds (floating point) since the epoch began. (In UNIX the
+    epoch started on January 1, 1970, 00:00:00 UTC)
+
+    Parameters
+    ----------
+    ipython : ``IPython.terminal.interactiveshell.TerminalInteractiveShell``
+        The active ``ipython`` ``Shell``, perhaps the one returned by
+        ``IPython.get_ipython()``.
+
+    Attributes
+    ----------
+    curr_time: float
+        The current time in seconds.
+
+    max_idle_time: int
+        The maximum number of seconds a user session can be idle (currently set
+        to 172800 seconds or 48 hours).
+
+    last_active_time: float
+        The time of the last user activity in this session.
+
+    idle_time: float
+        The amount of time the user session has been idle.
+    '''
+
+    def __init__(self, ipython):
+        self.curr_time = 0
+        self.max_idle_time = max_idle_time
+        self.last_active_time = 0
+        self.idle_time = 0
+
+        # _set_last_active_time() function will trigger every time user runs a cell
+        ipython.events.register('pre_run_cell', self._set_last_active_time)
+
+    def _set_last_active_time(self, result):
+        self.last_active_time = time.time()
+
+    def _get_time_passed(self):
+        self.curr_time = time.time()
+        self.idle_time = self.curr_time - self.last_active_time
+
+    def _timer(self, sleep_time):
+        time.sleep(sleep_time)
+
+    def _start_session(self):
+        # Check if idle_time has exceeded max_idle_time
+        while (self.idle_time < self.max_idle_time):
+            self._timer(self.max_idle_time - self.idle_time)
+            self._get_time_passed()
+
+        # Close the user session
+        print("This hutch-python session has timed out and automatically closed. Any code that is running will continue to run until it is completed.")
+
+        # Close this ipython session
+        get_ipython().ask_exit()
```

Comment:
We apparently need to follow this up with another exit call to kill the prompt that's waiting for input, otherwise as you say in the desc this waits for user input before exiting.
```
ip = get_ipython()
ip.ask_exit()
ip.pt_app.app.exit()
```

Reply:
`ip.pt_app.app.exit()` seems to generate an error message when the user code is running. I wasn't able to find another clean way to kill the prompt and exit ipython.

From what I observed the 'post_run_cell' hook triggers after the user enters something at the prompt. It doesn't trigger when the user's code is running. So, the self.last_active_time variable is set to the time that I entered something at the prompt and not when my user code is running.

In the screenshot 'post_run_cell' gets a timestamp of when I first pressed enter at the prompt (for example, at time 00:00:00). Then the timer starts to count down. After 20 seconds the timer script checks self.last_active_time, which still shows 00:00:00, so it ends the session. So, the session timeout message shows up 20 seconds after I entered code at the prompt and not 20 seconds after my user code finishes running.

<img width="880" alt="screenshot_of_exit_call" src="https://github.com/pcdshub/hutch-python/assets/168043200/9de8570a-bc00-4659-b541-c01ae47cbdf8">

Reply:
Ahh I think this error is because the `ip.pt_app.app` object is only "running" while waiting for user input

Reply:
That is to say, my suggestion here is only valid for cancelling the prompt input itself (automatically) and is only valid when there is a prompt.

Reply:
Yep, it worked well for me when there was no user code running.

Reply:
I think this is why I thought about using zeromq or websockets to check the status of the ipython kernel and figure out if it's busy or idle. If the kernel is idle, then we'd be able to wait for the user code to finish running and ip.pt_app.app would get ipython to exit cleanly.

Reply:
Maybe I'm also misunderstanding things, but it seems to me that the `post_run_cell` runs after the cell is finished executing 
```ipython
In [4]: for i in range(5):
   ...:     time.sleep(1)
   ...:     print(i)
   ...:
pre-run
0
1
2
3
4
post-run
```

So we could switch off the timer at `pre_run_cell`, then record the last active time and start counting down on the `post_run_cell`?

Reply:
From my observation `post_run_cell` is triggering after the first line `for i in range(5):`. When post_run_cell triggers it calls the `_set_active_time()` function, which sets `self.last_active_time` and the 20 second timer begins. When 20 seconds have passed the script checks how long the user has been idle (user has not entered anything at the prompt). If  self.last_active_time has not changed in the last 20 seconds the script prints out the "This hutch-python session has timed out..." message.

Adding to the confusion is that typing `time.sleep(1)` and `print(i)` doesn't seem to trigger `post_cell_run`. So the "This hutch-python session has timed out..." message is printed to the console just a few seconds after this loop runs.

I think my implementation might be a little convoluted / hard to read.

Reply:
So the snippet I posted was generated with this small modification to your code:
```python
        # _set_last_active_time() function will trigger every time user runs a cell
        ipython.events.register('pre_run_cell', self._set_last_active_time)
        ipython.events.register('pre_run_cell', lambda: print('pre-run'))
        ipython.events.register('post_run_cell', lambda: print('post-run'))
```

Which seems to conflict with your observation.  As currently written, this PR only registers an event to the `pre_run_cell` event right?

Reply:
Right, this PR only registers an event to the  'pre_run_cell' hook. I was testing with 'post_run_cell' to see how it behaves. 

I think there might be an issue with the logic in my code.  Going to track down the bug.

Reply:
I figured out the error in the script and rewrote the code so it's simpler. `ip.pt_app.app.exit()` should be working now. The test code has a 30 second timer and is in my development folder /cds/home/j/janeliu/git/hutch-python/hutch_python.